tests/test_api.py

```diff
@@ -314,3 +326,155 @@ def f():
+    assert b.to_scalar(m[1]) == 51.0
+
+
+def test__convert_arg_init_AbstractClass(backend_opt):
+    backend, backend_options = backend_opt
+    b = load_backend(backend, backend_options)
+
+    @dataclass(frozen=True)
+    class A():
+
+        s: 'scalar number'
+
+        def apply(self, input):
+            """Apply the layer."""
+            return (input, self.s)
+
+    model = A(2.0)
+    m = to_device(model, backend, backend_options)
+
+    assert isinstance(m, A)
+    assert isinstance(b.to_scalar(m.s), float)
+    assert b.to_scalar(m.s) == 2.0
+
+
+def test__convert_arg_init_AbstractArray(backend_opt):
+    backend, backend_options = backend_opt
+    b = load_backend(backend, backend_options)
+    m = to_device(MA(2, 3), backend, backend_options)
+
+    assert isinstance(m, ArrayWrapper)
+    np.testing.assert_allclose(b.to_numpy(m.array), MA(2, 3))
+
+
+def test__convert_arg_init_AbstractUnion(backend_opt):
+    # TODO
+    pass
+
+
+def test__convert_arg_init_AbstractScalar(backend_opt):
+    backend, backend_options = backend_opt
+    b = load_backend(backend, backend_options)
+
+    model1 = 3.0
+    m1 = to_device(model1, backend, backend_options)
+    assert isinstance(b.to_scalar(m1), float)
+    assert b.to_scalar(m1) == 3.0
+
+    model2 = 14
+    m2 = to_device(model2, backend, backend_options)
+    assert isinstance(b.to_scalar(m2), int)
+    assert b.to_scalar(m2) == 14
+
+
+def test_convert_arg_init(backend_opt):
+
+    backend, backend_options = backend_opt
+
+    def _convert(data, typ):
+        return _convert_arg_init(data,
+                                 to_abstract_test(typ),
+                                 load_backend(backend, backend_options))
+
+    with pytest.raises(TypeError):
+        _convert(newenv, EnvType)
+
+    # Class TypeError check and Tuple TypeError check
+
+    pt3 = Point3D(1, 2, 3)
+
+    with pytest.raises(TypeError):
+        _convert((1, 2), Point(i64, i64))
+    with pytest.raises(TypeError):
+        _convert(Point(i64, i64), (1, 2))
+
+    # Arrays
+
+    fmat = np.ones((5, 8), dtype='float64')
+    imat = np.ones((5, 8), dtype='int32')
+
+    with pytest.raises(TypeError):
+        _convert(imat, af64_of(5, 8))
+    with pytest.raises(TypeError):
+        _convert(fmat, ai32_of(5, 8))
+
+    # Misc errors
+
+    pt3 = Point3D(1, 2, 3)
+    with pytest.raises(TypeError):
+        _convert(10, f64)
+    with pytest.raises(TypeError):
+        _convert("blah", to_abstract_test("blah"))
+    with pytest.raises(TypeError):
+        _convert(1.5, i64)
+    with pytest.raises(TypeError):
+        _convert(10, (i64, i64))
+    with pytest.raises(TypeError):
+        _convert((1,), (i64, i64))
+    with pytest.raises(TypeError):
+        _convert((1, 2, 3), (i64, i64))
+    with pytest.raises(TypeError):
+        _convert((1, 2, 3), [i64])
+    with pytest.raises(TypeError):
+        _convert(pt3, Point(i64, i64))
+    with pytest.raises(TypeError):
+        _convert(1, Bool)
```

Comment:
These tests seem weird to me because _convert_arg_init() should only be used for to_device(), which can't really do meaningful type checking.  So why test that type checking is done?

Reply:
@abergeron 
I added these tests in order cover all the
```
if not isinstance(*, *):
            raise TypeError(f'Expected *')
```
in _convert_arg_init().

Do you want me to remove all the
```
if not isinstance(*, *):
            raise TypeError(f'Expected *')
```
in _convert_arg_init().

Reply:
Well you did just remove them so I guess that's it.